client/src/librarySymbolManager.ts

```diff
@@ -0,0 +1,174 @@
+        const files: vscode.Uri[] = (await Promise.all(fileDiscoveryActions)).flat();
+
+        for (const fileUri of files) {
+            symbolFileActions.push(this.processSymbolFile(fileUri));
+        }
+
+        // Process all symbol files, filtering out any that failed to load.
+        const allSymbolFiles: ReadonlyArray<[vscode.Uri, LibraryJson]> = (await Promise.all(symbolFileActions)).filter(
+            (value: [vscode.Uri, LibraryJson] | undefined) => value !== undefined,
+        ) as ReadonlyArray<[vscode.Uri, LibraryJson]>;
+
+        const validSymbolLibraries: Map<string, LibraryJson> = new Map<string, LibraryJson>();
+
+        for (const [uri, library] of allSymbolFiles) {
+            const moduleName: string = LibrarySymbolManager.getModuleNameFromFileUri(uri);
+            validSymbolLibraries.set(moduleName, library);
+        }
+
+        this.clientTrace?.info(`Registering symbol files. Total file count: ${validSymbolLibraries.size}`);
+
+        await this.librarySymbolClient
+            .setLibrarySymbols(validSymbolLibraries)
+            .then(() => this.registeredSymbolModules.push(...validSymbolLibraries.keys()));
+
+        return this.registeredSymbolModules;
+    }
+
+    public async getSymbolFilesFromDirectory(directory: vscode.Uri): Promise<vscode.Uri[]> {
+        let isDirectoryValid: boolean = false;
+
+        try {
+            const stat: vscode.FileStat = await this.fs.stat(directory);
+
+            if (stat.type !== vscode.FileType.Directory) {
+                this.clientTrace?.error(
+                    `${LibrarySymbolManager.ErrorMessagePrefix} '${directory.toString()}' is not a directory.`,
+                    JSON.stringify(stat),
+                );
+            } else {
+                isDirectoryValid = true;
+            }
+        } catch (error) {
+            this.clientTrace?.error(
+                `${LibrarySymbolManager.ErrorMessagePrefix} Exception while processing '${directory.toString()}'.`,
+                error,
+            );
+        }
+
+        if (!isDirectoryValid) {
+            return [];
+        }
+
+        const files: [string, vscode.FileType][] = await this.fs.readDirectory(directory);
+
+        // We only want .json files.
+        return files
+            .map((value: [string, vscode.FileType]): vscode.Uri | undefined => {
+                const fileName: string = value[0];
+
+                if (
+                    value[1] === vscode.FileType.File &&
+                    fileName.toLocaleLowerCase().endsWith(LibrarySymbolManager.SymbolFileExtension)
+                ) {
+                    return vscode.Uri.joinPath(directory, fileName);
+                }
+
+                return undefined;
+            })
+            .filter((value: vscode.Uri | undefined) => value !== undefined) as vscode.Uri[];
+    }
+
+    public async processSymbolFile(fileUri: vscode.Uri): Promise<[vscode.Uri, LibraryJson] | undefined> {
+        try {
```

Comment:
Do we need the `try` around the whole block, or just the library assert?

Reply:
I think we want the whole block since the `readFile` operation could throw.